### ISSUE TITLE
Hooks for file validation

### ISSUE THREAD

**Issue by prschmid** [jeffwidman/flask-uploads#18]
Right now files are "validated" based on their file extension. However, I can easily upload an unsupported file by simply changing the extension. E.g. ``somefile.sh`` to ``somefile.jpg``. So what would be useful is a way to add hooks to actually inspect the contents of the file to ensure that it is the type of file it claims to be. For example, if we were to use decorators, something like:
```python
@flask_uploads.inspect('jpg')
def inspect_jpg(f):
   """Inspect f to ensure that it is a JPG file"""
   data = open(f,'rb').read(11)
   if data[:4] != '\xff\xd8\xff\xe0': return False
   if data[6:] != 'JFIF\0': return False
   return True

@flask_uploads.inspect(['xls', 'xlsx'])
def inspect_excel(f): 
   """Inspect f to ensure that it is an Excel file"""
   import xlrd
   try:
      # Note, xlrd will close the book automatically... but this will read
      # the contents of this file into memory if it succeeds...
      xlrd.open_workbook(f)
   except:
     raise UploadNotAllowed()
```
Some of these inspectors could be built in and just enabled (e.g. JPG, since that is standard) and others could be left up to the user of the library (e.g. Excel).

This might be best added as subtype of a generic post ``save()`` hook. E.g.
```python
@flask_uploads.post_save
def after_upload(f):
   """Do something after the file has been saved to disk"""
   # Submit this file to be scanned by an anti-virus agent via a celery task
   perform_virus_scan.delay(f)
```
I'd leave it up to you as to which hooks come first (the generic post-save ones or the file inspection ones) as I could make an argument going either way. However my guess is that by default it should then ``os.remove()`` the file if there are any exceptions that occur in these hooks (since it's better to be safe and not have that file on disk).

**Comment by jeffwidman** [comment]
After looking at what you described, I completely agree there's a lot of utility in adding these hooks. Essentially it's a primitive event system with some default events for file validation.

What if we support three events?

1. `pre_save` - user-defined actions triggered before flask-uploads modifies the file. Custom file inspection should happen here.
2. <don't_know_what_to_call_it> - flask-uploads default events. Users can't hook in here. The two defaults that come to mind are:
 - file validation of common filetypes
 - checking if file already exists and resolving conflicts (#6)
3. `post_save` - user-defined actions triggered after flask-uploads finishes. Antivirus or other background jobs should hook in here.

It's important that actions hooked into the same event are executed in the order they were registered. I'm not sure where this info should be stored, as I'm still new to maintaining a flask extension, but I'm sure there's a way. Possibly it's just stored on the app object.

A couple of other related topics:
* Should we bundle in default file validators for common filetypes?
Yes, enabling validation by default makes sense, but only for filetypes where we have 100% confidence that a file isn't matching the file format. I haven't worked with files much at the byte level, but I suspect there's a lot of filetypes that may be difficult to validate 100% because the file format is a bit undefined. For other validation, we'll leave it up to the user to write their own action.

* What should be done with the file when an exception happens?
I think for custom decorators this should be somewhat undefined and let the user deal with it. Pre-save will raise an exception preventing the file from being saved unless the user catches the exception. Post save the exception won't delete the file unless the user catches the exception and deletes it.

* What should the API look like for the built-in events? 
I'm not sure. Enabling/disabling file validation per file extension can be a little complicated since they don't map 1:1 to upload sets. Resolving conflicts similarly needs some decisions made on what options to offer the user. I think overall on these questions though we could punt while adding the event system itself. 

Thoughts?

**Comment by prschmid** [comment]
Sorry for the delay, work/life got in the way of fun. =)

Yeah the ``pre_save`` and ``post_save`` hooks sound great. (We have hooks like that much of our production code here at work). For the the second item above, where you say "Users can't hook in here.", do you mean that users won't be able to make custom validators or just that there are no ``pre_validate``, ``post_validate`` hooks? The problem with making those things "unchangeable" is different people want to handle things differently. E.g. what happens if someone's view of what a valid XYZ file is not what the default is (e.g. only support versions 2+ of some file type). Maybe we are saying the same thing, but I'd suggest making some "built-in" hooks for these special events that are part of the more generic ``pre_save`` and ``post_save`` routine. That way you don't need to have special case code for these things. E.g.

```python
@flask_uploads.pre_save()
def first_thing_to_check(f):
  pass

@flask_uploads.pre_save()
def second_thing_to_check(f):
  pass

@flask_uploads.verify('jpg')
def verify_jpg(f):
  pass

@flask_uploads.verify(['xls', 'xlsx'])
def verify_jpg(f):
  pass

@flask_uploads.post_save()
def first_thing_post_save(f):
  pass
```
where the ``verify`` decorator is merely syntactic sugar around a ``pre_save`` decorator to make it easier for someone reading the code to understand the intent (and calls the right one based on the uploaded files extension). I.e. you could achieve the same thing with a single pre-save event (please excuse the pseudocode)
```python
@flask_uplaods.pre_save()
def validate(f):
  if f.extension == 'jpg':
     #do stuff
  if extension in ['xls', 'xlsx']:
     # do stuff
```
But since that is a bit of a pain to handle, the ``verify`` decorator just lets people generate those mappings in a more convenient way so that you only have to write the one pre-save event with a for loop around all of the custom ones that got registered. Then in the actual code for the uploading would merely be something like:
```python
for pre_save_fn in pre_save_fns:
  pre_save_fn(f)

# save the file to disk

for post_save_fn in post_save_fns:
  post_save_fn(f)
```
For file name conflicts, I would argue that this could be a special case that doesn't quite fall into these pre post- save hooks. I.e. I don't really think of it as a pre-save step, but more of a error handling in the event that there is a duplicate file. As always, different people want different things. On our prod system, for example, all files on disk are given UUIDs to ensure no collisions. How do you ensure that your re-naming/failing policy is what users want? As such, I'd suggest putting in a ``on_conflict`` "error" handler that can be used to do the right thing. E.g
```python
@flask_uploads.on_conflict
def on_filename_conflict(f):
  pass
```
and then in the uploading code actually call it
```python
for pre_save_fn in pre_save_fns:
  pre_save_fn(f)

on_conflict(f)
# save the file to disk

for post_save_fn in post_save_fns:
  post_save_fn(f)
```
Alternatively, if it truly is to be error handling you could do something like
```python
try:
  # save the file to disk
except ConflictError:
  on_conflict()
```
This might be harder/uglier since you'll probably want to save the file so you'll have to re-attempt the save after that error handler.

Regarding the questions above.

 * Should we bundle in default file validators for common filetypes?
I would say that this is the work of another branch of code. This branch is to provide the event hooks, and then make more custom validators later (and let people decide what they want). Also, not everyone will want to use this functionality -- maybe they already have a separate process that handles all of this.

* What should be done with the file when an exception happens?
Your points make sense. However, I would argue if that is the case, then I would add in another event ``before_return`` (or something like that with a better name) that gets called no matter what happens. I.e. basically the equivalent to a ``finally`` in a try-except block. This is where users would then be instructed to put any and all cleanup code. One stop shop.

* What should the API look like for the built-in events? 
I don't think the file validation for different file types is that hard do to. Since ``verify`` is just another ``pre_save`` event you can just handle them like any other event.

Welcome any feedback!

**Comment by jeffwidman** [comment]
FYI--I saw this, just haven't had time to sit down and think through it yet.

**Comment by prschmid** [comment]
I haven't gotten around to doing more on this either, but looks like something could be done using filemagic (https://pypi.python.org/pypi/filemagic/) if it's running on a *nix machine

**Comment by jeffwidman** [comment]
Agreed. Afraid I probably won't have time to implement this myself though, as I'm not using this library on any projects right now. 

If you're interested, I'm happy to add you as a collaborator on the project.

**Comment by shuttle1987** [comment]
Something like this:

```
@flask_uploads.verify('zip')
def verify_zip(f):
  pass
```

would be useful for me right now as I'm currently doing some things in a more ad-hoc manner like this in a project:

```
def create_from_zip(zippedFile):
    if zippedFile.mimetype != 'application/zip':                                                      
        logger.info("Non zip mimetype from request, got {}".format(zippedFile.mimetype))              
        return "File type must be zip", 415 
    if not zipfile.is_zipfile(zippedFile):
        logger.info("Zip file corrupted")
        return "File type must be zip", 415
```

And I think it would be cleaner to have this inside
